pkg/mutation/schema/schema.go

```diff
@@ -2,383 +2,179 @@ package schema
+// a string to a path and another might assign a list.
 func (db *DB) Upsert(mutator MutatorWithSchema) error {
+	if mutator == nil {
+		// This would indicate a bug in our code.
+		return ErrNilMutator
+	}
 	db.mutex.Lock()
 	defer db.mutex.Unlock()
-	return db.upsert(mutator, true)
+	return db.upsert(mutator)
 }
 
-func (db *DB) upsert(mutator MutatorWithSchema, unwind bool) error {
-	oldMutator, ok := db.mutators[mutator.ID()]
-	// unwind will be set to false if we are actually unwinding a bad commit
-	if ok && !oldMutator.HasDiff(mutator) && unwind {
-		return nil
-	}
-	if ok && unwind {
-		db.remove(oldMutator.ID())
+func (db *DB) upsert(mutator MutatorWithSchema) error {
+	id := mutator.ID()
+	if oldMutator, ok := db.cachedMutators[id]; ok {
+		if !mutator.HasDiff(oldMutator) {
+			// We've already added a Mutator which has the same path and bindings, so
+			// there's nothing to do.
+			return nil
+		}
+		db.remove(id)
 	}
 
-	modified := []*scheme{}
-	for _, gvk := range getSortedGVKs(mutator.SchemaBindings()) {
+	path := mutator.Path()
+	bindings := mutator.SchemaBindings()
+	db.cachedMutators[id] = mutator.DeepCopy().(MutatorWithSchema)
+
+	var conflicts references
+	for _, gvk := range bindings {
 		s, ok := db.schemas[gvk]
 		if !ok {
-			s = &scheme{gvk: gvk}
+			s = &node{}
 			db.schemas[gvk] = s
 		}
-		if err := s.add(mutator.Path().Nodes); err != nil {
-			// avoid infinite recursion
-			if unwind {
-				db.unwind(mutator, oldMutator, modified)
-			}
-			return err
-		}
-		modified = append(modified, s)
+		newConflicts := s.Add(id, path.Nodes)
+		conflicts = merge(conflicts, newConflicts)
 	}
-	m := mutator.DeepCopy()
-	db.mutators[mutator.ID()] = m.(MutatorWithSchema)
-	return nil
-}
 
-// unwind a bad commit
-func (db *DB) unwind(new, old MutatorWithSchema, schemes []*scheme) {
-	for _, s := range schemes {
-		s.remove(new.Path().Nodes)
-		if s.root == nil {
-			delete(db.schemas, s.gvk)
-		}
-	}
-	if old == nil {
-		return
+	for c := range conflicts {
+		db.conflicts[c] = true
 	}
-	if err := db.upsert(old, false); err != nil {
-		// We removed all changes made by the previous mutator and
-		// are re-adding a mutator that was already present. Because
-		// this mutator was already present and we have a lock on the
-		// db, this should never fail. If it does we are in an unknown
-		// state and should panic so we can recover by bootstrapping
-		// and raise the visibility of the issue.
-		log.Error(err, "could not upsert previously existing mutator into schema, this is not recoverable")
-		panic(err)
+	if len(conflicts) > 0 {
+		// Adding this mutator had schema conflicts with another, so return an error.
+		return fmt.Errorf("%w: the following Mutators have conflicting schemas: %s",
+			ErrConflictingSchema, conflicts.String())
 	}
+	return nil
 }
 
-// Remove removes the mutator with the given id from the
-// db.
+// Remove removes the mutator with the given id from the db.
 func (db *DB) Remove(id types.ID) {
```

Comment:
We need to figure out how to send a signal to the mutator controllers that mutators that no longer have conflicts need to be reconciled to update their status... ditto for mutators with conflicts.

Reply:
Per meeting, I've changed System to manually remove the offending mutator on error. This preserves the existing behavior and is now validated via unit test.